### PR TITLE
trace_tcpconnect: Fix missing src and dst ip and port on error

### PR DESCRIPTION
**Pull request by burak-ok**
When an error happens in the connect call the src ip addr + port and the dst port is cleared in the kernel before returning. Therefore we need to save all the information at a point when it is available.

Edit: Actual source line is: https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/af_inet.c#L738
~~See https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L350~~

## Before
```bash
RUNTIME.CONTAINERNAME        SRC                                   DST                                   COMM                  PID        TID        UID        GID ERROR     
quizzical_cannon             172.17.0.2:34286                      142.250.185.131:80                    wget                41798      41798          0          0           
quizzical_cannon             172.17.0.2:35018                      142.250.186.99:80                     wget                41798      41798          0          0           
quizzical_cannon             0.0.0.0:0                             192.168.2.1:0                         wget                41882      41882          0          0 ECONNREFUS
quizzical_cannon             0.0.0.0:0                             192.168.1.50:0                        wget                42026      42026          0          0 EHOSTUNREA
```

## After
```bash
RUNTIME.CONTAINERNAME        SRC                                   DST                                   COMM                  PID        TID        UID        GID ERROR     
quizzical_cannon             172.17.0.2:53208                      142.250.186.99:80                     wget                40138      40138          0          0           
quizzical_cannon             172.17.0.2:56386                      216.58.206.67:80                      wget                40138      40138          0          0           
quizzical_cannon             172.17.0.2:47944                      192.168.2.1:80                        wget                40195      40195          0          0 ECONNREFUS
quizzical_cannon             172.17.0.2:32952                      192.168.1.50:80                       wget                40736      40736          0          0 EHOSTUNREA
```

### CODE REVIEW

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (blanquicet) [inline gadgets/trace_tcpconnect/program.bpf.c:425]: Why not to use `kprobe/tcp_v{4|6}_connect`? It seems to me that the src and dst information is already present in the args:

- `sk` already contains the source IP and port: https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L218
- `uaddr` contains the dst IP and port: https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L216

Using `kprobe/tcp_v{4|6}_connect` will allow us to still get the [src IP/port](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L350) and [dst port](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L352) when there's an error sending the SYN (which is not covered with this PR).

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (blanquicet) [inline gadgets/trace_tcpconnect/program.bpf.c:453]: IMO, the logic became quite complex. I suggest creating a mermaid diagram as we did for [fsnotify](https://www.inspektor-gadget.io/docs/latest/gadgets/fsnotify#architecture) where we describe what we do at each hook. In particular, how programs interact with maps. WDYT?

File: gadgets/trace_tcpconnect/README.mdx
Comment (blanquicet) [inline gadgets/trace_tcpconnect/README.mdx:36]: Improvement for a follow-up PR:

Use here the example we had in the built-in version here: https://www.inspektor-gadget.io/docs/latest/gadgets/builtin/trace/tcpconnect#on-kubernetes (After updating it with new error field 🙂)

**Review by blanquicet — COMMENTED** [review]
Thanks for fixing it and creating an initial guide!

File: gadgets/trace_tcpconnect/README.mdx
Comment (burak-ok) [reply gadgets/trace_tcpconnect/README.mdx:106]: Even better: I didn't save the file before commiting

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (burak-ok) [reply gadgets/trace_tcpconnect/program.bpf.c:260]: It can happen that src/dst is already there. For example when calling `connect` on an already connected socket. I don't know for `SS_CONNECTING` though.

In general any other path/case than https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/af_inet.c#L654

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (burak-ok) [reply gadgets/trace_tcpconnect/program.bpf.c:425]: For example the source port is still unknown at the start of the function. It is set in https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L299 There `inet_hash_connect` is called which sets the port. See comment https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/inet_hashtables.c#L1164

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (blanquicet) [reply gadgets/trace_tcpconnect/program.bpf.c:260]: You're right. Could you please add a comment to clarify it? Something like:

```suggestion
	} else {
		// src_dst_entry is not set when tcp_v{4|6}_connect is not called, e.g.,
		// when calling inet_stream_connect on an already connected socket. So,
		// try to read the endpoints from the socket here.
		read_l4endpoints_from_sock_v4(&event->src, &event->dst, sk);
	}
```

Same in `trace_v6` and `exit_inet_stream_connect`.

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (blanquicet) [reply gadgets/trace_tcpconnect/program.bpf.c:454]: It'd be highly appreciated 🙂

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (burak-ok) [reply gadgets/trace_tcpconnect/program.bpf.c:426]: I don't understand what you are suggesting, maybe I missed something.

This `kretprobe/tcp_v4_connect` was introduced because we don't have the port and ip in `exit_inet_stream_connect` when an error happened

Comment (blanquicet) [reply gadgets/trace_tcpconnect/program.bpf.c:426]: I'm suggesting anticipating the collection of the src/dst ip and port from `kretprobe/tcp_v{4|6}_connect` to `kprobe/tcp_v{4|6}_connect` because by using `kretprobe/tcp_v{4|6}_connect`, we'll not provide the src IP/port and dst port when there's an error in `tcp_v{4|6}_connect` (while sending the SYN) because they are cleaned up: [here](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L350) and [here](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L353), respectively.

The problem with applying my suggestion is that we will not provide the src port if there is an error in `inet_stream_connect` because, as you said, the src port is still unknown in `kprobe/tcp_v{4|6}_connect`. So, let's continue with your proposal, but please document what is still missing.

Comment (burak-ok) [reply gadgets/trace_tcpconnect/program.bpf.c:426]: We could maybe trace `inet_hash_connect` instead then: https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L299
But I am currently not so sure what this function does right after the hash connect: https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L305

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (burak-ok) [reply gadgets/trace_tcpconnect/program.bpf.c:454]: Could you please take a look. Since I am currently too familiar with the topic I want to make sure its understandable for everyone

File: gadgets/trace_tcpconnect/README.mdx
Comment (alban) [inline gadgets/trace_tcpconnect/README.mdx:147]: Do you mean in the kretprobe? Or from the kprobe?

```suggestion
        Note right of inet_stream_connect: Gather all information in the kretprobe<br/>and submit the event
```

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (alban) [inline gadgets/trace_tcpconnect/program.bpf.c:84]: Is it true with synchronous sockets? Could the thread block in the `tcp_v4_connect` function, and if we have more than max_entries=32 threads on the system blocked on that function, the map could get full.

Comment (burak-ok) [reply gadgets/trace_tcpconnect/program.bpf.c:84]: That is also true. What size would you suggest?

Comment (burak-ok) [reply gadgets/trace_tcpconnect/program.bpf.c]: For now I increased it from `32` to `128`

File: gadgets/trace_tcpconnect/README.mdx
Comment (alban) [inline gadgets/trace_tcpconnect/README.mdx:33]: Note that by removing the flags sections, it will not be readded automatically when new flags are added in the gadget because the sed command would not find it:
https://github.com/inspektor-gadget/inspektor-gadget/blob/main/gadgets/Makefile#L79

In my opinion, it is fine to keep the flags section. Or the Makefile could be updated.

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (alban) [inline gadgets/trace_tcpconnect/program.bpf.c:87]: I don't understand why max_entries is different between the two maps `sockets_per_process` and `src_dst_per_process`.

Comment (burak-ok) [reply gadgets/trace_tcpconnect/program.bpf.c:87]: It might be a (premature) optimization.

`sockets_per_process` stores a pointer, whereas `src_dst_per_process` is at least 42 bytes large. `src_dst_per_process` gets written into after the `SYN` package gets sent, but with synchronous sockets the waiting is still done afterwards.
So we might need to have the larger buffer?

File: gadgets/trace_tcpconnect/README.mdx
Comment (alban) [inline gadgets/trace_tcpconnect/README.mdx:132]: s/skb/sock
It is not a socket buffer (skb) but a socket.

File: gadgets/trace_tcpconnect/program.bpf.c
Comment (alban) [inline gadgets/trace_tcpconnect/program.bpf.c:263]: The logic does not work in some edge-cases. But it was already not working before.

Steps to reproduce:

* Start the gadget:
```
$ sudo -E ig run ghcr.io/inspektor-gadget/gadget/trace_tcpconnect:burak-trace_tcpconnect-fix --verify-image=false
WARN[0000] image signature verification is disabled due to using corresponding option 
WARN[0000] image signature verification is disabled due to using corresponding option 
RUNTIME.CONTAINERNAME                 SRC                                               DST                                              COMM                    PID        TID ERROR         
```

* Start a container. To make it reproduce faster, change the tcp timeout of the container to approximately 7 seconds:
```
$ docker run -ti --rm --privileged busybox
/ # sysctl net.ipv4.tcp_syn_retries=2
net.ipv4.tcp_syn_retries = 2
```

* Start a tcp connection to nowhere and quickly stop the process with ctrl-z (SIGSTOP), and resume it with `fg`:
```
/ # wget 1.2.3.4
Connecting to 1.2.3.4 (1.2.3.4:80)
^Z[1]+  Stopped                    wget 1.2.3.4
/ # fg
```

* Notice the gadget output after waiting a few seconds more:
```
RUNTIME.CONTAINERNAME  SRC               DST         COMM  PID     TID ERROR
boring_edison          172.17.0.3:32798  1.2.3.4:80  wget  434444  434444 error#512
boring_edison          0.0.0.0:0         1.2.3.4:0   wget  434444  434444 ETIMEDOUT
```

There is two events (one emitted at the ctrl-z, and one emitted at the end of the tcp timeout).

**Review by alban — APPROVED** [review]
LGTM, thanks.
Please fix s/skb/sock before merging.
IMO the edge case problem  https://github.com/inspektor-gadget/inspektor-gadget/pull/3517#discussion_r1822990039 can be fixed in a later PR.

**Comment by alban** [comment]
I think `tracepoint/sched/sched_process_exit` should clean up `src_dst_per_process` and `sockets_per_process` in case `kretprobe/inet_stream_connect` does not get called due to maxactive being too low (example of [kretprobe-maxactive issue](https://github.com/iovisor/bcc/issues/1072)).

**Pull request merged**
